website/content/docs/architecture/scale.mdx

```diff
@@ -0,0 +1,283 @@
+
+When the leader takes snapshot B at index 199, it truncates the logs that accumulated between snapshot A and snapshot B, which means it truncates Raft logs with indexes between 100 and 199. 
+
+Because the new server restored snapshot A, the new server has a current index of 99. It requests logs 100 to 150 because index 150 was the current index when it started the replication restore process. At this point, the leader recognizes that it only has logs 200 and higher, and does not have logs for indexes 100 to 150. The leader determines that the new server’s state is stale and starts the process over by sending the new server the latest snapshot, snapshot B.  
+
+Consul keeps a configurable number of [Raft trailing logs](/consul/docs/agent/config/config-files#raft_trailing_logs) to prevent the snapshot install loop from repeating. The trailing logs are the last logs that went into the snapshot, and the new server can more easily catch up to the current state using these logs. The default Raft trailing logs configuration value is suitable for most deployments.
+
+In Consul v1.10 and higher, operators can try to prevent a snapshot install loop by monitoring and comparing Consul servers’ `consul.raft.rpc.installSnapshot` and `consul.raft.leader.oldestLogAge` timing metrics. Monitor these metrics for the following situations:
+
+- After truncation, the lowest number on `consul.raft.leader.oldestLogAge` should always be at least two times higher than the lowest number for `consul.raft.rpc.installSnapshot`.
+- If these metrics are too close, increase the number of Raft trailing logs, which increases `consul.raft.leader.oldestLogAge`. Do not set the Raft trailing logs higher than necessary, as it can negatively affect write throughput and latency.
+
+For more information, refer to [Raft Replication Capacity Issues](h/consul/docs/agent/telemetry#raft-replication-capacity-issues).
+
+## Performance considerations for specific use cases
+
+This section provides configuration and monitoring recommendations for Consul deployments according to the features you prioritize and their use cases.
+
+### Service discovery
+
+To optimize performance for service discovery, we recommend deploying multiple small clusters with consistent numbers of service instances and watches.
+
+Several factors influence Consul performance at scale when used primarily for its service discovery and health check features. The factors you have control over include: 
+
+- The overall number of registered service instances
+- The use of [stale reads](/consul/api-docs/features/consistency#consul-dns-queries) for DNS queries
+- Max number of service instances for a given service  - Are there services that are registered in every node or in a large percentage of nodes in the fleet?  
```

Comment:
I don't recall who added this consideration, but I don't _think_ it was me. We may have to follow up with that person to figure out the answer to your question.

Reply:
@jmurret : Do you know who might know the answer to Jeff's question?

Reply:
I think it's a bad idea to include something that we can't answer ourselves. It raises more questions for the audience (and us) than it answers.

If this were originally written in markdown in Git, we'd have an easy way to find out who wrote this line :) But because it was written in Google docs, I'm not sure how to track down who wrote this.

I think we should either:
- Figure out who originally wrote this (and what it means), OR
- Remove it, with the understanding that we can update the doc later to include this if necessary

Reply:
@preetapan : Do you happen to know who wrote this line or what it's referring to? Neither John nor I remember writing it, and we're unable to personally provide clarification here.

Reply:
@preetapan : Connecting your reply back to Jeff's original questions:

> My unresolved question: What specifically is influencing performance? Is it too many instances registered for a given service, or the dispersal of instances for a single service across nodes?

If I understand correctly, this third bullet is not about performance impacts on the Consul server nodes, but instead on the client/dataplane nodes hosting the workloads?

Let's say I have 100 instances of service `foo` on 100 nodes. One instance of that service has a health check that moves from `passing` to `warning`. Does that mean every node with a service `foo` is affected by that update to 1 service instance?

If there are no client agents, only dataplane agents, does the max number of service instances for a given service still matter? (i.e., Do I need to consider it in the specifications for my Consul server nodes?)

Reply:
> Does that mean every node with a service foo is affected by that update to 1 service instance?

Any entity that is monitoring the catalog for changes to this service will receive an update from the servers about state changes related to registered instances. The entity could a Consul agent, the data plane component, an API client, etc. Regardless of the type, the receiving entity will need to process that state change and reconcile it with the previously known data for the service. There's a computational cost associated with this.

> Do I need to consider it in the specifications for my Consul server nodes?

I think so. It's also important to consider how these update are being dispersed to consumers of the catalog info. AFAIK, it's computationally more expensive to distribute these updates to an HTTP API consumer a part of blocking query response as opposed to disseminating the information over Consul's streaming [g]RPC system, or xDS.

Reply:
would the following suggestion be accurate?

```suggestion
- The number of entities that are monitoring Consul (for example: Consul agent, data plane component, API client, etc...) for changes for a service. When the Consul server sends updates about a service's state, the receiving entity needs to process the state change and reconcile it with the previously known data for the service. 
```

Reply:
nit: I'd use "dataplane" rather than "data plane", since that's how we refer to the `consul-dataplane` component elsewhere.

Slightly altered wording. I'll let @blake make the final call on accuracy.

```suggestion
- The number of entities that are monitoring Consul, such as Consul client agents or dataplane components, for changes in a service's instances including registration and health status. When any service change occurs, all those entities incur a computational cost because they must process the state change and reconcile it with previously known data for the service.
```

Reply:
my feedback on the above is that its technically acccurate but now has lost the original point about the "fanout" factor i.e number of unique instances of the service. How about this? 

-The number of instances that a specific service has across your fleet manifests as additional computational load. This is proportional to the the number of entities that are monitoring Consul, such as Consul client agents or dataplane components. When any change occurs in a service's state, all those entities incur a computational cost because they must process the state change and reconcile it with previously known data for the service.